docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,319 @@
+relation to the constraints they have set and in relation with other
+considerations - specifically replication factor and replica diversity. This can
+lead to real problems: it’s unclear when a whole locality can be turned down
+without losing quorum for any ranges.  
+As a result of this work, admins will also be able to script actions to be taken
+when partitioning, or another zone config change, has been “fully applied”.
+
+
+# Guide-level explanation
+
+## Background
+
+Reminder for a few CRDB concepts:
+
+**Replication zone configs and constraints:** key spans can be grouped into
+*“replication zones” and properties can be set for each zone through a “zone
+*config”. Among others, a zone config specifies the replication factor for the
+*data in the zone (i.e. the replication factor of all the ranges in the zone), a
+*list of constraints, and a list of lease preference policies.  
+**Replication Constraints:** used to restrict the set of nodes/stores that can
+*hold a replica for the respective ranges. Constraints can say that only nodes
+*marked with certain localities or attributes are to be used or that certain
+*nodes/stores have to be avoided. Constraints can also be specified at the level
+*of a replica, not only a range; for example, constraints can be set to require
+*that some ranges have one replica in locality A, one in B and one in C.  
+**Lease preferences:** Lease preferences are expressed just like constraints, except
+they only affect the choice of a leaseholder. For example, one can say that the
+leaseholder needs to always be in a specific locality. Unlike constraints, not
+being able to satisfy a lease preference doesn’t stop a lease acquisition.  
+**Replica diversity:** Given a set of constraints (or no constraints), CRDB tries to
+split replicas evenly between localities/sublocalities that match the
+constraint.  
+
+## Constraint conformance
+
+Currently the only insight an administrator has into the health of replication
+is via the Admin UI which displays a counter for under-replicated ranges and one
+for unavailable ranges.
+There’s nothing about constraints, lease preferences or diversity. Also there’s
+no information on over-replicated ranges - although hopefully those will go away
+once we get atomic group membership changes. Also, the way in which these
+counters are computed is defective: the counters are incoherent and, for
+example, if all the nodes holding a partition go away, the ranges in that
+partition are not counted (see
+https://github.com/cockroachdb/cockroach/issues/19644).
+
+Besides the general need for an administrator to know that everything is
+copacetic, there are some specific cases where the lack of information is
+particularly unfortunate:
+
+
+1. Testing fault tolerance: A scenario people ran into was testing CRDB’s
+resiliency to networking failures. A cluster is configured with 3 Availability
+Zones (AZs) and the data is constrained to require a replica in every AZ (or
```

Comment:
I assume we will apply this to single machine, data center/availability zone, and region failures as all are important for testing here

Reply:
i just thought it would be worth explaining that it applies to all but we can remove that if you think its unnecessary